src/briefcase/platforms/macOS/__init__.py

```diff
@@ -82,10 +82,92 @@ class macOSMixin:
     # 0.3.20 introduced a framework-based support package.
     platform_target_version = "0.3.20"
 
+    def is_icloud_synced(self, path: Path) -> bool:
+        """Determine if a path is on an iCloud drive.
+
+        This is done by looking for the "com.apple.fileprovider.fpfs#P" resource fork.
+        This fork only appears on *some* directories - most notably, `.app` folders.
+
+        :param path: The location to check.
+        :returns: True if the location has iCloud resource markers.
+        """
+        # Check if the path is on an iCloud mounted drive.
+        try:
+            # Check for the iCloud resource fork. "Good" operation produces an error,
+            # so use quiet mode.
+            self.tools.subprocess.check_output(
+                [
+                    "xattr",
+                    "-p",
+                    "com.apple.fileprovider.fpfs#P",
+                    path,
+                ],
+                quiet=1,
+            )
+            # The resource fork was found.
+            return True
+        except subprocess.CalledProcessError:
+            # The resource fork was not found.
+            # This includes the file not existing.
+            return False
+
+    def verify_not_on_icloud(self, app: AppConfig, cleanup=False):
+        """Confirm that the app is *not* on an iCloud synchronized drive.
+
+        When a `.app` folder is on an iCloud-synchronized drive, iCloud adds filesystem
+        metadata to the folder. This metadata can't be removed (iCloud will just put it
+        back again), but it also conflicts with app signing. So - if we detect this
+        metadata, the project has been generated somewhere that ultimately won't work.
+
+        Optionally, this method will clean up the bundle if the verification fails.
+
+        :param app: The app to check.
+        :param cleanup: Should the app bundle be deleted if verification fails?
+        """
+        if self.is_icloud_synced(self.binary_path(app)):
+            msg = [
+                "Briefcase cannot be used on an iCloud-mounted drive "
+                "(such as your Documents folder)."
```

Comment:
This message could be confusing if the user's Documents folder isn't shared on iCloud, but their Briefcase project is.

Reply:
Good point; I've updated the exception messages.